### PR TITLE
[Merged by Bors] - `Commands` benchmarking

### PR DESCRIPTION
**Pull request by NathanSWard**
# Objective

- Currently the `Commands` and `CommandQueue` have no performance testing. 
- As `Commands` are quite expensive due to the `Box<dyn Command>` allocated for each command, there should be perf tests for implementations that attempt to improve the performance.  

## Solution

- Add some benchmarking for `Commands` and `CommandQueue`.

### CODE REVIEW

**Comment by NathanSWard** [comment]
I'm not too happy with these current benchmarks. 
However this was split apart from #2332 to reduce the scope of that PR.
If you have suggestions on better perf tests for commands, please let me know!

**Comment by TheRawMeatball** [comment]
A seeded RNG for optionally adding these components would be useful imo

File: benches/benches/bevy_ecs/commands.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    group.measurement_time(std::time::Duration::from_secs(5));
+
+    for i in (1..5).map(|i| i * 2) {
+        let entity_count = i * 1000;
+
+        group.bench_function(format!("{}_entities", entity_count), |bencher| {
+            let mut world = World::default();
+            let mut command_queue = CommandQueue::default();
+
+            let mut rng = StdRng::seed_from_u64(42);
```
Comment (TheRawMeatball) [inline benches/benches/bevy_ecs/commands.rs:70]: We use see if using SmallRng is more performant to minimize the impact of the randomizer in benchmark.

**Comment by mockersf** [comment]
Looking at the benches, it feels like they test more entity spawning and component insertion than commands themselves.

Would it be interesting to add a dummy command that doesn't do much and use it in a bench?

**Comment by NathanSWard** [comment]
> Looking at the benches, it feels like they test more entity spawning and component insertion than commands themselves.
Would it be interesting to add a dummy command that doesn't do much and use it in a bench?

Yep, I like this idea :)
I went ahead and added benchmarks for `FakeCommands`

**Review by mockersf — APPROVED** [review]

**Comment by mockersf** [comment]
to show even more the impact of the size of the command, you could try with `struct MyCommand<T>` and run bench inserting one of:
* `MyCommand<()>` typical empty command like spawning an entity
* `MyCommand<(u32, u32, u32)>` typical small command like inserting a component
* `MyCommand<LargeStruct>` typical large command like inserting a bundle

**Comment by NathanSWard** [comment]
> to show even more the impact of the size of the command, you could try with struct FakeCommand<T> {field: T} and run benches for each of:

Would the idea be to run benchmarks where you add only `FakeCommand<()>` N times to the command queue.
(In order words you don't mix different sized commands?)

**Comment by NathanSWard** [comment]
> FakeCommand<LargeStruct> typical large command like inserting a bundle

Also by `LargeStruct` how _large_ are we talking?

**Comment by mockersf** [comment]
> Would the idea be to run benchmarks where you add only `FakeCommand<()>` N times to the command queue.

... probably? 😄 

> Also by `LargeStruct` how large are we talking?

I think you picked a good value 👍 

My idea was to explore how your other pr on commands would impact various sized commands, looks like you found something interesting with 12 bytes commands!

**Comment by NathanSWard** [comment]
Ideally I'd like to merge this first, then merge #2332.
However I'll have to change one or the other because the `Command` trait got change from `self: Box<Self>` to `self`

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
RNG has overhead that I'm worried will get in the way of showing the actual cost of these operations. The goal of rng here is to produce "different archetype shapes", but we could almost as easily hard code these shapes, which would also give us the benefit of reproducibility.

**Comment by NathanSWard** [comment]
> RNG has overhead that I'm worried will get in the way of showing the actual cost of these operations. The goal of rng here is to produce "different archetype shapes", but we could almost as easily hard code these shapes, which would also give us the benefit of reproducibility.

Hopefully with a seeded-rng this would lessen the impact of rng, however, it is still rng soo... :shrug:
It's not so much to make archetype shapes, but rather to have the `CommandQueue` not know what type of command will be pushed to it. 

However, we can achieve similar results with just a modulo inside a `black_box` so we should all be good :)
I went ahead and removed the rng.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**